pkg/controller/endpointslice/metrics/cache.go

```diff
@@ -0,0 +1,118 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"math"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// NewCache returns a new Cache with the specified endpointsPerSlice.
+func NewCache(endpointsPerSlice int32) *Cache {
+	return &Cache{
+		maxEndpointsPerSlice: endpointsPerSlice,
+		globalCache:          map[types.NamespacedName]ServicePortCache{},
+	}
+}
+
+// Cache tracks values for total numbers of desired endpoints as well as the
+// efficiency of EndpointSlice endpoints distribution.
+type Cache struct {
+	lock                 sync.Mutex
+	maxEndpointsPerSlice int32
+	numEndpoints         int
+	globalCache          map[types.NamespacedName]ServicePortCache
+}
+
+// ServicePortCache tracks values for total numbers of desired endpoints as well
+// as the efficiency of EndpointSlice endpoints distribution for each unique
+// Service Port combination.
+type ServicePortCache map[endpointutil.PortMapKey]EfficiencyInfo
+
+// EfficiencyInfo stores NumEndpoints and NumSlices for calculating total
+// numbers of desired endpoints and the efficiency of EndpointSlice endpoints
+// distribution.
+type EfficiencyInfo struct {
+	NumEndpoints int
+	NumSlices    int
+}
+
+// UpdateServicePortCache updates a ServicePortCache in the global cache for a
+// given Service and updates the corresponding metrics.
+func (c *Cache) UpdateServicePortCache(serviceNN types.NamespacedName, spCache ServicePortCache) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	prevNumEndpoints := 0
+	if existingSPCache, ok := c.globalCache[serviceNN]; ok {
+		prevNumEndpoints = numEndpoints(existingSPCache)
+	}
+
+	currNumEndpoints := numEndpoints(spCache)
+	c.numEndpoints = c.numEndpoints + currNumEndpoints - prevNumEndpoints
+
+	c.globalCache[serviceNN] = spCache
+	c.updateMetrics()
+}
+
+// DeleteService removes references of a Service from the global cache and
+// updates the corresponding metrics.
+func (c *Cache) DeleteService(serviceNN types.NamespacedName) {
+	if spCache, ok := c.globalCache[serviceNN]; ok {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		c.numEndpoints = c.numEndpoints - numEndpoints(spCache)
+		c.updateMetrics()
+		delete(c.globalCache, serviceNN)
+	}
+}
+
+func (c *Cache) efficiency() float64 {
+	actualSlices := 0
+	desiredSlices := 0
+	for _, spCache := range c.globalCache {
+		for _, eInfo := range spCache {
+			actualSlices += eInfo.NumSlices
+			desiredSlices += numDesiredSlices(eInfo.NumEndpoints, int(c.maxEndpointsPerSlice))
+		}
+	}
+	return float64(desiredSlices) / float64(actualSlices)
```

Comment:
An idea: how about just record number of desiredSlices and actualSlices as separate number? 

efficiency can be calculated on query? that way it will be easier to change efficiency calculation and lower the load on controller side to precalculate this.

Reply:
Good call, I've switched to using 2 separate gauge metrics here. I think that will allow for more advanced use cases and requires a bit less controller code.